File: verifiers/go/internal/audit/audit.go
// Package audit enforces the ADR-002 invariant at the runner ingress
// boundary: the executor's reasoning trace must never reach the
// verifier model — or the verifier tools. The daemon already runs
// this guard on its side; we re-run it here because (a) the daemon
// could change without the runner noticing, (b) defence in depth is
// cheaper than the brand-existential cost of a leaked
// chain-of-thought, and (c) the runner is the last place we control
// before the diff hits real tooling.
//
// The guard works on a generic decoded JSON map (the same blob the
// runner just unmarshalled from stdin). Any field whose lower-case
// name contains a denylist substring causes a Refusal. Recursive on
// nested maps and arrays-of-maps.
package audit

import (
	"fmt"
	"sort"
	"strings"
)

// Denylist is the lower-case substring set the guard checks against
// each JSON field name. Kept in sync with
// apps/verifier/internal/verification/verification.go.
//
// We bias toward false-positives. A legitimate field that trips this
// guard must be renamed, not whitelisted: the cost of one rename is
// trivial vs. the cost of accidentally piping reasoning through.
var Denylist = []string{
	"reasoning",
	"chain_of_thought",
	"chain-of-thought",
	"cot",
	"thinking_trace",
	"thinking-trace",
	"thoughts",
	"scratchpad",
	"internal_monologue",
	"hidden_state",
	"agent_trace",
	"executor_trace",
	"trajectory",
	"plan_critique",
	"reflection",
}

// LeakageError reports the first offending field and the substring it
// matched. The runner surfaces this as a tool_unavailable TestReport
// and refuses to run any tier.
type LeakageError struct {
	OffendingField string
	Pattern        string
}

func (e *LeakageError) Error() string {
	return fmt.Sprintf("audit: executor-reasoning leak — field %q matched denylist pattern %q (ADR-002 invariant)",
		e.OffendingField, e.Pattern)
}

// NoLeakage walks an arbitrary decoded-JSON value and returns a
// LeakageError on the first denylisted field. Pass the parsed
// stdin map[string]any directly.
func NoLeakage(payload map[string]any) error {
	return walkMap(payload, "")
}

func walkMap(m map[string]any, prefix string) error {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	// Sort so the offending field reported is deterministic across runs.
	sort.Strings(keys)
	for _, k := range keys {
		full := k
		if prefix != "" {
			full = prefix + "." + k
		}
		lk := strings.ToLower(k)
		for _, deny := range Denylist {
			if strings.Contains(lk, deny) {
				return &LeakageError{OffendingField: full, Pattern: deny}
			}
		}
		switch v := m[k].(type) {
		case map[string]any:
			if err := walkMap(v, full); err != nil {
				return err
			}
		case []any:
			for i, elem := range v {
				if mm, ok := elem.(map[string]any); ok {
					if err := walkMap(mm, fmt.Sprintf("%s[%d]", full, i)); err != nil {
						return err
					}
				}
			}
		}
	}
	return nil
}

File: verifiers/python/crucible_verify_python/audit.py
"""Defence-in-depth audit guard against executor-reasoning leakage.

The Go side already audits the same denylist at request ingest
(see ``apps/verifier/internal/verification/verification.go``::``auditMap``).
This module re-applies the same rules inside the Python runner so a
mis-routed payload cannot reach a tool subprocess that might serialise it
back into the report.

Mirror the Go list exactly — drift between the two is a load-bearing bug.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from typing import Any

# Intentionally aggressive — false positives are cheaper than the
# brand-existential cost of a leaked reasoning trace (ADR-002).
REASONING_DENYLIST: tuple[str, ...] = (
    "reasoning",
    "chain_of_thought",
    "chain-of-thought",
    "cot",
    "thinking_trace",
    "thinking-trace",
    "thoughts",
    "scratchpad",
    "internal_monologue",
    "hidden_state",
    "agent_trace",
    "executor_trace",
    "trajectory",
    "plan_critique",
    "reflection",
)

# Path patterns banned inside Diff.Files[*].Path.
REASONING_PATH_PATTERNS: tuple[str, ...] = (
    ".reasoning.",
    "/reasoning/",
    ".cot.",
    "/cot/",
    "_thinking_",
    "_scratchpad_",
    "agent_trace",
    "executor_trace",
)


@dataclass(frozen=True, slots=True)
class LeakageError(Exception):
    """Raised when a payload contains a reasoning-tagged field."""

    field: str
    pattern: str

    def __str__(self) -> str:  # pragma: no cover — trivial
        return (
            f"executor-reasoning leak detected — field {self.field!r} "
            f"matched pattern {self.pattern!r} (ADR-002 invariant)"
        )


def _match_key(key: str) -> str | None:
    lowered = key.lower()
    for deny in REASONING_DENYLIST:
        if deny in lowered:
            return deny
    return None


def audit_payload(payload: Any, prefix: str = "") -> None:
    """Recursively scan ``payload`` for reasoning-tagged keys.

    Accepts arbitrary JSON-decoded values (dicts, lists, scalars). Raises
    :class:`LeakageError` on the first match — we fail closed.
    """
    if isinstance(payload, Mapping):
        # Sort to make audit errors deterministic across Python dict orders.
        for key in sorted(payload.keys(), key=str):
            full = f"{prefix}.{key}" if prefix else str(key)
            match = _match_key(str(key))
            if match is not None:
                raise LeakageError(field=full, pattern=match)
            audit_payload(payload[key], full)
        return
    if isinstance(payload, (list, tuple)):
        # str/bytes don't satisfy this branch (they're not list/tuple
        # subclasses), so we don't need to filter them out explicitly.
        for i, item in enumerate(payload):
            audit_payload(item, f"{prefix}[{i}]")
        return
    # Scalars (str/int/float/bool/None) are not audited by value — that
    # would produce too many false positives on natural English. We only
    # gate on field names, matching the Go side.


def audit_diff_paths(paths: Iterable[str]) -> None:
    """Refuse paths that obviously came from a reasoning sidecar.

    Mirrors ``isReasoningPath`` in the Go verification package.
    """
    for path in paths:
        lowered = path.lower()
        for pattern in REASONING_PATH_PATTERNS:
            if pattern in lowered:
                raise LeakageError(field=f"diff.files.{path}", pattern="path-pattern")
